src/vcpkg/sourceparagraph.cpp

```diff
@@ -708,142 +787,280 @@ namespace vcpkg
     };
     ContactsDeserializer ContactsDeserializer::instance;
 
-    static constexpr StringLiteral EXPRESSION_WORDS[] = {
-        "WITH",
-        "AND",
-        "OR",
-    };
-    static constexpr StringLiteral VALID_LICENSES[] =
-#include "spdx-licenses.inc"
-        ;
-    static constexpr StringLiteral VALID_EXCEPTIONS[] =
+    static constexpr StringLiteral VALID_LICENSES[] = {
 #include "spdx-licenses.inc"
-        ;
+    };
+    static constexpr StringLiteral VALID_EXCEPTIONS[] = {
+#include "spdx-exceptions.inc"
+    };
 
-    // We "parse" this so that we can add actual license parsing at some point in the future
-    // without breaking anyone
-    struct LicenseExpressionDeserializer : Json::IDeserializer<std::string>
+    // The "license" field; either:
+    // * a string, which must be an SPDX license expression.
+    //   EBNF located at: https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#license
+    // * `null`, for when the license of the package cannot be described by an SPDX expression
+    static bool in_list_case_insensitive(StringView id, View<StringLiteral> lst)
     {
-        virtual StringView type_name() const override { return "an SPDX license expression"; }
+        return lst.end() !=
+               Util::find_if(lst, [id](StringLiteral el) { return Strings::case_insensitive_ascii_equals(id, el); });
+    }
 
-        enum class Mode
+    ExpectedT<std::string, msg::LocalizedString> parse_spdx_license_expression(StringView sv, LicenseIdCheck check)
+    {
+        enum class Expecting
         {
-            ExpectExpression,
-            ExpectContinue,
-            ExpectException,
+            License,        // at the beginning, or after a compound (AND, OR)
+            Exception,      // after a WITH
+            CompoundOrWith, // after a license
+            Compound,       // after an exception (only one WITH is allowed), or after a close paren
         };
 
-        virtual Optional<std::string> visit_string(Json::Reader&, StringView sv) override
+        if (sv.empty())
         {
-            Mode mode = Mode::ExpectExpression;
-            size_t open_parens = 0;
-            std::string current_word;
+            return msg::format(msgEmptyLicenseExpression);
+        }
+
+        auto next_token = [sv, current_word = StringView{sv.begin(), sv.begin()}]() mutable
+            -> ExpectedT<StringView, msg::LocalizedString> {
+            auto is_whitespace = Parse::ParserBase::is_whitespace;
+            auto is_idstring_element = [](char ch) { return ch == '.' || Parse::ParserBase::is_alphanumdash(ch); };
```

Comment:
I don't see any reason for this to be a nested lambda like this. Should probably just be a nonmember (along with `is_whitespace` and `is_alphanumdash`).